kernel/src/hil/time.rs

```diff
@@ -81,20 +90,96 @@ pub trait Alarm: Time {
 
     /// Returns the value set in [`set_alarm`](#tymethod.set_alarm)
     fn get_alarm(&self) -> u32;
+
+    /// Set the client for interrupt events.
+    fn set_client(&'a self, client: &'a AlarmClient);
+
+    /// Returns whether this alarm is currently active (will eventually trigger
+    /// a callback if there is a client).
+    fn is_enabled(&self) -> bool;
+
+    /// Enables the alarm using the previously set `tics` for the alarm.
+    ///
+    /// Most implementations should use the default implementation which calls `set_alarm` with the
+    /// value returned by `get_alarm` unless there is a more efficient way to achieve the same
+    /// semantics.
+    fn enable(&self) {
+        self.set_alarm(self.get_alarm())
+    }
+
+    /// Disables the alarm.
+    ///
+    /// The implementation will _always_ disable the alarm, however, it may be possible for an
+    /// alarm to have already expired but the event not delivered to the client. In this case, the
+    /// implementation must return [`FAIL`] letting the caller know that an event for the alarm
+    /// will still be delivered.
+    fn disable(&self) -> ReturnCode;
 }
 
-/// A client of an implementor of the [`Alarm`](trait.Alarm.html) trait.
-pub trait Client {
+/// A client of an implementer of the [`Alarm`](trait.Alarm.html) trait.
+pub trait AlarmClient {
     /// Callback signaled when the alarm's clock reaches the value set in
     /// [`Alarm#set_alarm`](trait.Alarm.html#tymethod.set_alarm).
     fn fired(&self);
 }
 
 /// The `Timer` trait models a timer that can notify when a particular interval
 /// has elapsed.
-pub trait Timer: Time {
+pub trait Timer<'a>: Time {
+    /// Set the client for interrupt events.
+    fn set_client(&'a self, client: &'a TimerClient);
+
     /// Sets a one-shot timer to fire in `interval` clock-tics.
+    ///
+    /// Calling this method will override any existing oneshot or repeating timer.
     fn oneshot(&self, interval: u32);
+
     /// Sets repeating timer to fire every `interval` clock-tics.
+    ///
+    /// Calling this method will override any existing oneshot or repeating timer.
     fn repeat(&self, interval: u32);
+
+    /// Returns the interval for a repeating timer.
+    ///
+    /// Returns `None` if the timer is disabled or in oneshot mode and `Some(interval)` if it is
+    /// repeating.
```

Comment:
I think we should change this to:

```rust
/// If the timer is enabled, returns `Some(interval)` if this is is a repeating timer or `None` if it is a oneshot timer.
/// If the timer is disabled, the return value is undefined.
```

It seems like an easy mistake for an implementation with a software interval to implement this as just `return self.interval` -- I'm not sure it makes sense for the implementer of this method to need to internally check whether the timer is enabled.